compiler/vm/vm.nim

```diff
@@ -1279,69 +1825,160 @@ proc rawExecute(c: PCtx, start: int, tos: PStackFrame): TFullReg =
 
         fn(outp)
     of opcContainsSet:
+      # a = c in b
       decodeBC(rkInt)
-      regs[ra].intVal = ord(inSet(regs[rb].node, regs[rc].regToNode))
+      let val = regs[rc].intVal
+      assert val >= 0 and val <= high(uint16).BiggestInt
+      regs[ra].intVal = ord(uint16(regs[rc].intVal) in regs[rb].atomVal.setVal.elems)
     of opcSubStr:
-      decodeBC(rkNode)
+      # a = substr(rb, rc, rd)
+      decodeBC(akString)
       inc pc
       assert c.code[pc].opcode == opcSubStr
       let rd = c.code[pc].regA
-      createStr regs[ra]
-      regs[ra].node.strVal = substr(regs[rb].node.strVal,
-                                    regs[rc].intVal.int, regs[rd].intVal.int)
+
+      # TODO: validate user input (the range) and report errors
+      # TODO: reuse data of regs[ra].strVal
+      regs[ra].strVal = substr(regs[rb].strVal, regs[rc].intVal.int, regs[rd].intVal.int, c.memory)
     of opcParseFloat:
+      # TODO: this op has really unusual semantics. Turn it into a callback?
+
+      # a = number of chars read
+      # rc[] = parseFloat(rb, rd)
       decodeBC(rkInt)
       inc pc
       assert c.code[pc].opcode == opcParseFloat
       let rd = c.code[pc].regA
+
       var rcAddr = addr(regs[rc])
-      if rcAddr.kind == rkRegisterAddr: rcAddr = rcAddr.regAddr
-      elif regs[rc].kind != rkFloat:
-        regs[rc] = TFullReg(kind: rkFloat)
-      regs[ra].intVal = parseBiggestFloat(regs[rb].node.strVal,
-                                          rcAddr.floatVal, regs[rd].intVal.int)
+      if rcAddr.kind == rkRegAddr: rcAddr = rcAddr.regAddr
+      else:
+        ensureKind(regs[rc], rkFloat, c.memory)
+
+      # TODO: don't do a string copy here
+      regs[ra].intVal = parseBiggestFloat($regs[rb].strVal,
+                                          rcAddr[].floatVal, regs[rd].intVal.int)
     of opcRangeChck:
+      # Checks if a is in range [b, c], aborts execution otherwise
       let rb = instr.regB
       let rc = instr.regC
-      if not (leValueConv(regs[rb].regToNode, regs[ra].regToNode) and
-              leValueConv(regs[ra].regToNode, regs[rc].regToNode)):
+
+      func isLessEqual(a, b: ptr TFullReg): bool =
+        ## Return whether a <= b with temporary conversion if required
+        
+        # This has the same behaviour as `semfold.leValueConv` (which was used previously) 
+
+        case a.kind
+        of rkInt:
+          case b.kind
+          of rkInt: a.intVal <= b.intVal
+          of rkFloat: a.intVal <= round(b.floatVal).int
+          else: false
+        of rkFloat:
+          case b.kind
+          of rkInt: a.floatVal <= toFloat64(toInt128(b.intVal))
+          of rkFloat: a.floatVal <= b.floatVal
+          else: false
+        else: false
+
+      let a = addr regs[ra]
+      let l = addr regs[rb]
+      let h = addr regs[rc]
+
+      if not (isLessEqual(l, a) and
+              isLessEqual(a, h)):
+        # TODO: are range errors `Exception`s or `Defect`s in normal Nim? If they
```

Comment:
So instead of aborting execution here, a guest catchable exception should be raised? The current behaviour (aborting) is inherited from the old code and is also what happens for the other back-ends